neuraxle/distributed/streaming.py

```diff
@@ -0,0 +1,416 @@
+        :class:`Observer`,
+        :class:`Observable`
+    """
+
+    def connect_queue_workers(self):
+        for i, (name, step) in enumerate(self):
+            if i != 0:
+                self[i - 1].subscribe(step)
+
+    def connect_queue_joiner(self, queue_joiner: 'QueueJoiner'):
+        self[-1].subscribe(queue_joiner)
+
+    def notify_new_batch_to_process(self, data_container_batch: DataContainer):
+        self[0].put(QueuedPipelineTask(step_name=self[0].name, data_container=data_container_batch))
+
+
+class ParallelQueuedPipeline(BaseQueuedPipeline):
+    """
+    Using :class:`QueueWorker`, run all steps in parallel using QueueWorkers.
+
+    .. seealso::
+        :class:`QueueWorker`,
+        :class:`BaseQueuedPipeline`,
+        :class:`SequentialQueuedPipeline`,
+        :class:`QueueJoiner`,
+        :class:`Observer`,
+        :class:`Observable`
+    """
+
+    def connect_queue_workers(self):
+        # nothing to do here, queue workers don't listen to each other in a ParallelQueuedPipeline
+        pass
+
+    def connect_queue_joiner(self, queue_joiner: 'QueueJoiner'):
+        for i, (name, step) in enumerate(self):
+            step.subscribe(queue_joiner)
+
+    def notify_new_batch_to_process(self, data_container_batch):
+        for i, (name, step) in enumerate(self):
+            step.put(QueuedPipelineTask(step_name=step.name, data_container=data_container_batch))
+
+
+class QueueJoiner(ObservableQueue):
+    """
+    Observe the results of the queue worker of type :class:`QueueWorker`.
+    Synchronize all of the workers together.
+
+    .. seealso::
+        :class:`QueuedPipeline`,
+        :class:`Observer`,
+        :class:`ListDataContainer`,
+        :class:`DataContainer`
+    """
+
+    def __init__(self, n_batches):
+        self.mutex_processing_in_progress = Lock()
+        self.mutex_processing_in_progress.acquire()
+        self.n_batches_left_to_do = n_batches
+        self.result = {}
+        ObservableQueue.__init__(self, Queue())
+
+    def join(self, original_data_container: DataContainer) -> DataContainer:
+        """
+        Return the accumulated results received by the on next method of this observer.
+
+        :return: transformed data container
+        :rtype: DataContainer
+        """
+        while self.n_batches_left_to_do > 0:
+            task: QueuedPipelineTask = self.queue.get()
+            self.n_batches_left_to_do -= 1
+
+            if task.step_name not in self.result:
+                self.result[task.step_name] = ListDataContainer(
+                    current_ids=[],
+                    data_inputs=[],
+                    expected_outputs=[],
+                    summary_id=None
+                )
+
+            self.result[task.step_name].concat(task.data_container)
+
+        data_containers = self._get_resulting_data_containers()
+        return original_data_container.set_data_inputs(data_containers)
+
+    def _get_resulting_data_containers(self):
```

Comment:
work in progress here in this function... I need to reorder the results by summary id